FIXES #194 - add loaded flag to lazy load property sources with parent_strategy

elrayle opened the pull request:
Fixes #194, #193

elrayle commented:
This isn't quite as pure a lazy load as I would like.  In further ad hoc testing, if a property with class_name defined has multiple values, then when you reference that property, a rdf_source will be created for each value.  This does still prevent explosion of the graph, but if you have a property like pages for a book with 1,000 pages, then a source object will be created for each of the 1,000 pages once the first page is referenced.

Why does this happen?

```ruby
page1 = book1.has_page.first
```
By referencing book1.has_page, it is accessing all values stored in has_page causing each page to be loaded.

elrayle commented:
Moved reload for lazy load to RDFSource.get_relation such that only the currently referenced resource will be reloaded.  All peers continue to remain unloaded until they are referenced.

mjsuhonos commented on lib/active_triples/rdf_source.rb:
```diff
@@ -106,6 +106,7 @@ def initialize(*args, &block)
       Array.wrap(self.class.type).each do |type|
         unless self.get_values(:type).include?(type)
           self.get_values(:type) << type
+          persistence_strategy.not_loaded if persistence_strategy.kind_of? ParentStrategy
```
Is this actually required?  I don't get any test failures if I remove it.  Similarly, I wonder if it should be further up the method like so:
```ruby
      unless args.first.is_a?(Hash) || args.empty?
        set_persistence_strategy(ParentStrategy)
        persistence_strategy.parent = args.shift
        persistence_strategy.not_loaded
      else
        set_persistence_strategy(RepositoryStrategy)
      end
```

mjsuhonos commented on lib/active_triples/persistence_strategies/parent_strategy.rb:
```diff
@@ -30,7 +30,21 @@ def persisted?
     ##
-    # Destroys the resource by removing it graph and references from the 
+    # Mark that this resource has not yet been loaded.
+    def not_loaded
```
:+1: mutators [should end in bangs](http://stackoverflow.com/questions/612189/why-are-exclamation-marks-used-in-ruby-methods).

> On Mar 10, 2016, at 2:21 PM, Thomas Johnson <notifications@github.com> wrote:
> 
> In lib/active_triples/persistence_strategies/parent_strategy.rb <https://github.com/ActiveTriples/ActiveTriples/pull/199#discussion_r55734333>:
> 
> > @@ -30,7 +30,21 @@ def persisted?
> >      end
> >  
> >      ##
> > -    # Destroys the resource by removing it graph and references from the 
> > +    # Mark that this resource has not yet been loaded.
> > +    def not_loaded
> Let's add a ! to the method sig: def not_loaded!
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/ActiveTriples/ActiveTriples/pull/199/files#r55734333>.
>

mjsuhonos replied:
:+1: mutators [should end in bangs](http://stackoverflow.com/questions/612189/why-are-exclamation-marks-used-in-ruby-methods).

elrayle commented:
Short answer:  Taking another look at the changes, I was able to simplify.  I removed not_loaded method and let loaded? default to false which removed the need to set it at all.  Testing loaded? for the first time will set loaded=false if it hasn't already been set to true by some other means.

Long answer:  At one point, the place I added the code to do the lazy load was causing resources with blank nodes to fail.  By setting the loaded flag when the type triple had to be added separate from what was read from the repository, it isolated the reload to only non-blanknode resources.  Some changes I made after that obviously fixed the issue and setting loaded in that way was no longer necessary.

Thanks for the reviews!!!  The simplification is better.

elrayle commented:
Good catch...  Setting loaded=true got deleted in some of the cleanup.

I added a test to make sure reload always loads from final_parent after first load.
